crates/flox-rust-sdk/src/models/publish.rs

```diff
@@ -356,22 +359,136 @@ impl PublishFlakeRef {
+                flake_ref: Some(FlakeRef::GitPath(file_ref.clone()).into()),
+                ..Default::default()
+            };
+            command
+                .run_typed(nix, &Default::default())
+                .await
+                .map_err(ConvertFlakeRefError::LocalFlakeMetadata)?
+        };
+
+        if local_metadata.revision.is_none() {
+            Err(ConvertFlakeRefError::LocalFlakeDirty)?;
+        }
+
+        // Create a handle to the git repo
+        let repo = Git::discover(file_ref.url.path())
+            .await
+            .map_err(|_| ConvertFlakeRefError::RepoNotFound(file_ref.url.path().into()))?;
+
+        // Get the upstream branch information.
+        // This is essentialy
+        //
+        //   upstream_ref = git rev-parse @{u}
+        //   (remote_name, branch_name) = split_once "/" upstream_ref
+        //   upstream_url = git remote get-url ${remote_name}
+        //   upstream_rev = git ls-remote ${remote_name} ${branch_name}
+        //
+        // The current branch MUST have an upstream ref configured
+        // which resolves to a valid rev on the remote.
+        let (remote_name, remote_url, remote_branch, remote_revision) = repo
+            .get_origin()
+            .await
+            .map_err(ConvertFlakeRefError::NoRemote)?;
+
+        // Ensure the remote branch exists
+        let remote_revision = remote_revision.ok_or(ConvertFlakeRefError::RemoteBranchNotFound)?;
+
+        info!("Resolved local flake to remote '{remote_name}:{remote_branch}' at '{remote_url}'");
+
+        // Check whether the local branch is in sync with its upstream branch,
+        // to ensure we publish the intended revision,
+        // by comparing the local revision to the one found upstream.
+        //
+        // Dirty branches are already permitted due to the filter above.
+        if let Some(local_rev) = local_metadata.revision {
+            if local_rev.as_ref() != remote_revision {
+                Err(ConvertFlakeRefError::RemoteBranchNotSync(
+                    local_rev.to_string(),
+                    remote_revision.clone(),
+                ))?
+            }
+        }
 
-    fn try_from(value: FlakeRef) -> Result<Self, Self::Error> {
-        let publish_flake_ref = match value {
+        // Git supports special urls for e.g. ssh repos, e.g.
+        //
+        //   git@github.com:flox/flox
+        //
+        // Normalize these urls to proper URLs for the use with nix.
+        let remote_url = git_url_parse::normalize_url(&remote_url)
+            .map_err(|e| ConvertFlakeRefError::UnknownRemoteUrl(e.to_string()))?;
+
+        // Copy the flakeref attributes but unlock it in support of preferred upstream refs
+        let mut attributes = file_ref.attributes;
+        attributes.last_modified = None;
+        attributes.rev_count = None;
+        attributes.nar_hash = None;
+        attributes.rev = Some(remote_revision.parse().unwrap());
```

Comment:
If you're going to unwrap something it's good practice to leave a comment stating that you know it can't fail, and it's probably better practice to use `expect` as well.